src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -491,5 +494,100 @@ private static CastResult TryGet(nuint source, nuint target)
 
             return ref Unbox_Helper(toTypeHnd, obj);
         }
+
+        internal struct ArrayElement
+        {
+            public object? Value;
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static ref object? ThrowArrayMismatchException()
+        {
+            throw new ArrayTypeMismatchException();
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        private static ref object? LdelemaRef(Array array, int index, void* type)
+        {
+            ArrayElement[] arr = Unsafe.As<ArrayElement[]>(array);
+
+            // this will throw appropriate exceptions if array is null or access is out of range.
+            void* elementType = RuntimeHelpers.GetMethodTable(arr)->ElementType;
+            ref object? element = ref arr[index].Value;
+
+            if (elementType != type)
+                goto throwArrayMismatch;
```

Comment:
I guess that's why we need `Unsafe.Unlikely()` (or Assume) 🙂

Reply:
The profile guide optimizations would take care of this as well, and not require people to think about it...

Reply:
Also, the JIT should see that `ThrowArrayMismatchException` always throws and do this reordering for you.  Is this goto really required?

Reply:
Moving the allocate/throw to a separate method seems important. Otherwise it adds more register use, push/pop, etc.
Reordering to the end - probably not. Just following the shape of original implementation. Will check.

Reply:
Separate method is fine. I was commenting on the use of goto.

Reply:
I've tried reordering the actual return and throw helper. JIT does not reorder them back. Not without PGO anyways. There is very tiny advantage from having the throwing call later in the method (1% or so, maybe even noise), so I will keep the order.  
`goto` is unnecessary either way, so I will remove it.